library/psa_crypto_slot_management.c

```diff
@@ -276,6 +277,66 @@ static psa_status_t psa_load_persistent_key_into_slot( psa_key_slot_t *slot )
 }
 #endif /* MBEDTLS_PSA_CRYPTO_STORAGE_C */
 
+#if defined(MBEDTLS_PSA_CRYPTO_BUILTIN_KEYS)
+
+static psa_status_t psa_load_builtin_key_into_slot( psa_key_slot_t *slot )
+{
+    psa_status_t status = PSA_ERROR_CORRUPTION_DETECTED;
+    psa_key_attributes_t attributes = PSA_KEY_ATTRIBUTES_INIT;
+    psa_key_lifetime_t lifetime = PSA_KEY_LIFETIME_VOLATILE;
+    psa_drv_slot_number_t slot_number = 0;
+    uint8_t *key_buffer = NULL;
+    size_t key_buffer_size = 0;
+    size_t key_buffer_length = 0;
+
+    /* Load keys in the 'builtin' range through their own interface */
+    if( ! psa_key_id_is_builtin(
+            MBEDTLS_SVC_KEY_ID_GET_KEY_ID( slot->attr.id ) ) )
+    {
+        return( PSA_ERROR_DOES_NOT_EXIST );
+    }
+
+    /* Check the platform function to see whether this key actually exists */
+    psa_set_key_id( &attributes, slot->attr.id );
+    status = mbedtls_psa_platform_get_builtin_key(
+                slot->attr.id, &lifetime, &slot_number );
+    if( status != PSA_SUCCESS )
+        return( status );
+
+    /* Set mapped lifetime on the attributes */
+    psa_set_key_lifetime( &attributes, lifetime );
+
+    /* If the key should exist according to the platform, load it through the
+     * driver interface. */
+    status = psa_driver_wrapper_get_key_buffer_size( &attributes,
+                                                     &key_buffer_size );
+    if( status != PSA_SUCCESS )
+        return( status );
+
+    key_buffer = mbedtls_calloc( 1, key_buffer_size );
+    if( key_buffer == NULL )
+        return( PSA_ERROR_INSUFFICIENT_MEMORY );
+
+    status = psa_driver_wrapper_get_builtin_key(
+                slot_number, &attributes,
+                key_buffer, key_buffer_size, &key_buffer_length );
+    if( status != PSA_SUCCESS )
+        goto exit;
+
+    status = psa_copy_key_material_into_slot(
+                slot, key_buffer, key_buffer_length );
+    if( status != PSA_SUCCESS )
+        goto exit;
+
+    /* Copy core attributes into the slot on success */
+    slot->attr = attributes.core;
+
+exit:
+    mbedtls_free( key_buffer );
```

Comment:
If the function returns in error the clean-up could be done by `psa_wipe_key_slot()` in `psa_get_and_lock_key_slot()` but we have to take care not returning PSA_ERROR_DOES_NOT_EXIST when a key buffer has been allocated. Thus all together probably safer to free the key buffer in the key slot here.

Reply:
This comment is saying that you agree with the current construction, yes?

Reply:
The code has changed here thus this is anyway out-dated. But otherwise yes I was kind of thinking loud here.